::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: Copyright (c) 2018 Mirantis Inc., Enea AB and others.
:
: All rights reserved. This program and the accompanying materials
: are made available under the terms of the Apache License, Version 2.0
: which accompanies this distribution, and is available at
: http://www.apache.org/licenses/LICENSE-2.0
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Wed, 22 Aug 2018 01:35:06 +0200
Subject: [PATCH] OPNFV package installation, Ubuntu user

* Install OpenSSH server (and client), so other OPNFV projects can
  easily connect to the Salt master node;
* Install 'salt-formula-gnocchi' distro package (empty git formula
  workaround);
* While at it, create 'ubuntu' user so other OPNFV projects don't
  have to switch to 'root' login;
* Preinstall `salt_minion_dependency_packages` and
  `salt_minion_reclass_dependencies`;
* Fetch opnfv.tar if OPNFV_EXTRA URI is set (to be used for injecting
  OPFNV formulas and reclass.system git repos during build);
* Use formula-fetch.sh from pending PR #38 [1] that allows local FS
  and/or Gerrit URIs for salt formula sources;

[1] https://github.com/salt-formulas/salt-formulas-scripts/pull/38

Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 DockerMake.yml | 44 +++++++++++++++++++++++++++++++++++++++++++-
 Pipfile        |  1 +
 2 files changed, 44 insertions(+), 1 deletion(-)

diff --git a/DockerMake.yml b/DockerMake.yml
index 2c75586..c6ea944 100644
--- a/DockerMake.yml
+++ b/DockerMake.yml
@@ -29,6 +29,7 @@ common-cleanup:

 common:
   build: |
+    ARG CACHE_INVALIDATE="0"
     RUN echo "Layer with common packages" \
       && eval ${LAYER_PKGUPDT} \
       && apt-get upgrade -qy \
@@ -102,12 +103,49 @@ salt-formulas:
     ENV SALT_ENV_PATH_ $SALT_ENV_PATH_
     ARG RECLASS_BASE="/srv/salt/reclass"
     ENV RECLASS_BASE $RECLASS_BASE
+    ARG OPNFV_EXTRA=""
+    ENV OPNFV_EXTRA $OPNFV_EXTRA
     RUN echo "Layer python/salt module prerequisites, formulas" \
+      && test -z "$OPNFV_EXTRA" || curl -sSqL $OPNFV_EXTRA | tar xf - -C /tmp \
       && mkdir -p /srv/salt \
-      && curl -sSqL https://raw.githubusercontent.com/salt-formulas/salt-formulas-scripts/master/formula-fetch.sh -o /srv/salt/formula-fetch.sh \
+      && curl -sSqL https://raw.githubusercontent.com/salt-formulas/salt-formulas-scripts/252ddc6/formula-fetch.sh -o /srv/salt/formula-fetch.sh \
       && bash -c 'source /srv/salt/formula-fetch.sh && setupPyEnv && fetchAll' \
       && eval ${LAYER_CLEANUP}

+opnfv:
+  build: |
+    # Some formulas have empty git repos, use the package version
+    ARG SALT_FORMULA_VERSION="nightly"
+    ENV SALT_FORMULA_VERSION $SALT_FORMULA_VERSION
+    ARG SALT_FORMULA_LIST="salt-formula-gnocchi"
+    ENV SALT_FORMULA_LIST $SALT_FORMULA_LIST
+    ENV APT_REPOSITORY "deb [arch=amd64] http://apt.mirantis.com/xenial ${SALT_FORMULA_VERSION} salt"
+    ENV APT_REPOSITORY_GPG "http://apt.mirantis.com/public.gpg"
+    RUN echo "Layer extra salt-formulas packages" \
+      && echo "$APT_REPOSITORY" | tee /etc/apt/sources.list.d/salt-formulas.list >/dev/null \
+      && curl -sL $APT_REPOSITORY_GPG | $SUDO apt-key add - \
+      && eval ${LAYER_PKGUPDT} \
+      && ${LAYER_INSTALL} ${SALT_FORMULA_LIST} -y --fix-missing -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" \
+      && eval ${LAYER_CLEANUP}
+    RUN echo "Layer with OPNFV packages" \
+      && eval ${LAYER_PKGUPDT} \
+      && ${LAYER_INSTALL} \
+           gawk \
+           inetutils-ping \
+           kmod \
+           net-tools \
+           openssh-server \
+           python-m2crypto \
+           python-msgpack \
+           python-netaddr \
+           python-oauth \
+           python-psutil \
+           python-yaml \
+      && useradd -m ubuntu \
+      && echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' > /etc/sudoers.d/ubuntu \
+      && eval ${LAYER_CLEANUP}
+
+
 wheel:
   requires:
     - base
@@ -138,9 +176,13 @@ reclass:
     ENV SALT_ENV_PATH_ $SALT_ENV_PATH_
     ARG RECLASS_BASE="/srv/salt/reclass"
     ENV RECLASS_BASE $RECLASS_BASE
+    ARG OPNFV_EXTRA=""
+    ENV OPNFV_EXTRA $OPNFV_EXTRA
     RUN echo "Layer reclass" \
+      && test -z "$OPNFV_EXTRA" || curl -sSqL $OPNFV_EXTRA | tar xf - -C /tmp \
       && mkdir -p /etc/reclass $RECLASS_BASE/classes/system \
       && git clone https://github.com/Mirantis/reclass-system-salt-model $RECLASS_BASE/classes/system \
+      && rm -rf /tmp/opnfv \
       && find $RECLASS_BASE/classes/system
     RUN pip install --install-option="--prefix=/usr/local" -I \
            "git+https://github.com/salt-formulas/reclass.git@$RECLASS_VERSION"
diff --git a/Pipfile b/Pipfile
index d3e8d66..6067dd2 100644
--- a/Pipfile
+++ b/Pipfile
@@ -9,6 +9,7 @@ name = "pypi"
 pygithub = "*"
 docker-make = {git = "https://github.com/avirshup/DockerMake"}
 dockermake = {git = "https://github.com/avirshup/DockerMake"}
+invoke = "*"

 [requires]
 python_version = "3.6"
